dashboard/modules/runtime_env/runtime_env_agent.py

```diff
@@ -102,17 +196,97 @@ def run_setup_with_logger():
                 per_job_logger.debug(f"Worker has resource :"
                                      f"{allocated_resource}")
                 context = RuntimeEnvContext(env_vars=runtime_env.env_vars())
-                self._pip_manager.setup(
-                    runtime_env, context, logger=per_job_logger)
-                self._conda_manager.setup(
-                    runtime_env, context, logger=per_job_logger)
-                self._py_modules_manager.setup(
-                    runtime_env, context, logger=per_job_logger)
-                self._working_dir_manager.setup(
-                    runtime_env, context, logger=per_job_logger)
                 self._container_manager.setup(
                     runtime_env, context, logger=per_job_logger)
 
+                # TODO(architkulkarni): Unify the below using Plugin class
+                # Set up working_dir
+                working_dir_uri = self._working_dir_manager.get_uri(
+                    runtime_env)
+                if working_dir_uri is not None:
+                    if working_dir_uri not in self._working_dir_uri_cache:
+                        size_bytes = self._working_dir_manager.create(
+                            working_dir_uri,
+                            runtime_env,
+                            context,
+                            logger=per_job_logger)
+                        self._working_dir_uri_cache.add(
+                            working_dir_uri, size_bytes, logger=per_job_logger)
+                    else:
+                        self._working_dir_uri_cache.mark_used(
+                            working_dir_uri, logger=per_job_logger)
+                    self._working_dir_manager.modify_context(
+                        working_dir_uri, runtime_env, context)
+
+                # Set up conda
+                conda_resources_dir = os.path.join(self._runtime_env_dir,
+                                                   "conda")
+                try_to_create_directory(conda_resources_dir)
+                conda_file_lock = os.path.join(self._runtime_env_dir,
+                                               "ray-conda-install.lock")
+                # TODO(architkulkarni): add comment
+                with FileLock(conda_file_lock):
```

Comment:
hmm this part smells really funny to me -- why is it necessary to do it here and not in the plugin class?

Reply:
I think we want (1) checking the URI cache, (2) installing the conda env and (3) updating the URI cache to all be under a single lock.  Otherwise we get a race condition when two workers both need the same conda env and see an empty URI cache.  But I'm a little shaky on the best practices for locking, maybe you know a better way?

Reply:
I see, I guess the alternative is to do this: 
Inside the plugin class, change `create_conda_env` to `create_conda_env_if_needed`, and only put a lock around the call `create_conda_env_if_needed`.  

I was originally concerned about how to decide the "if_needed" part, because if you just check the existence of the conda env path, you can't tell whether the env is fully installed or if it's still in progress.  But if the call to `create_conda_env_if_needed` is wrapped in a lock, then only one process can access it at a time, so we won't have this problem.